PublicMapping/districtbuilder
# Refactor server to move CPU intensive tasks to worker threads

## Overview

Per [ADR 5](https://github.com/PublicMapping/districtbuilder/blob/develop/doc/arch/adr-05-server-concurrency-strategy.md), we need to move our CPU intensive operations off of the main-thread.

As such, in this PR we:
 - Move `merge`, `import`, & `export` into worker threads using [`threads.js`](https://threads.js.org/)
   - I also moved `getTopologyProperties` to a worker, but that was more for memory concerns than CPU use
 - We now cache `topo.buf` files on disk so we can load them as needed in worker threads
 - Add caching in each worker using `lru-cache`
   - I have a PR open for updating the `lru-cache` types: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58813, until then I also have the types added to `global.d.ts`
   - I'm using [`object-sizeof`](https://www.npmjs.com/package/object-sizeof) to determine the approximate size of each topology layer
 - Route requests to workers based on which worker was most recently used for the region of the request
 
### Checklist


### Demo
N/A

### Notes

Along the way I also found and fixed an issue with duplicating maps for archived regions: https://github.com/PublicMapping/districtbuilder/commit/7ecb73116ba6ad30257b7c507a4283274bc63ee9

## Testing Instructions

- `scripts/update`
- `scripts/server`
- The application should function as normal:
  - Import
  - Export
  - Editing a map
- Restart your server, and watch as it loads _much_ faster 🎉

Closes #1138
